gooddata-sdk/gooddata_sdk/catalog/data_source/model/data_source.py

```diff
@@ -0,0 +1,183 @@
+            if self._URL_VENDOR:
+                db_vendor = self._URL_VENDOR
+            elif self._DATA_SOURCE_TYPE:
+                db_vendor = self._DATA_SOURCE_TYPE.lower()
+            return (
+                self._URL_TMPL.format(
+                    **self.custom_attributes.attributes,
+                    # url contains {db_vendor}, e.g. jdbc:{db_vendor}://....
+                    # we inject custom or default (DS_TYPE.lower()) value there
+                    db_vendor=db_vendor,
+                )
+                + self._join_params(";")
+            )
+        else:
+            raise Exception("Neither url(constructor) nor URL_TMPL set, cannot setup final url.")
+
+    def _join_params(self, delimiter: str) -> str:
+        if self.url_params:
+            return delimiter.join([p[0] + "=" + p[1] for p in self.url_params])
+        return ""
+
+    def _make_data_source_type(self) -> str:
+        if self._DATA_SOURCE_TYPE:
+            return self._DATA_SOURCE_TYPE
+        else:
+            raise Exception(
+                "Neither data_source_type(constructor) nor DATA_SOURCE_TYPE(class var) set, "
+                + "cannot setup final data_source_type"
+            )
+
+    @classmethod
+    def from_api(cls, entity: dict[str, Any]) -> CatalogDataSource:
+        ea = entity["attributes"]
+        credentials = Credentials.create(cls._SUPPORTED_CREDENTIALS, entity)
+        return cls(
+            id=entity["id"],
+            name=ea["name"],
+            data_source_type=ea["type"],
+            url=ea["url"],
+            schema=ea["schema"],
+            credentials=credentials,
+            enable_caching=ea.get("enableCaching"),
+            cache_path=ea.get("cachePath"),
+        )
+
+    def to_api(self) -> JsonApiDataSourceInDocument:
+        kwargs = self.credentials.to_api_args()
+        if self.enable_caching is not None:
+            kwargs["enableCaching"] = self.enable_caching
+        if self.cache_path is not None:
+            kwargs["cachePath"] = self.cache_path
+        return JsonApiDataSourceInDocument(
+            data=JsonApiDataSourceIn(
+                id=self.id,
+                attributes=JsonApiDataSourceInAttributes(
+                    name=self.name,
+                    type=self.data_source_type,
+                    url=self.url,
+                    schema=self.schema,
+                    **kwargs,
+                ),
+            )
+        )
+
+    @classmethod
+    def to_api_patch(cls, data_source_id: str, attributes: dict) -> JsonApiDataSourcePatchDocument:
+        return JsonApiDataSourcePatchDocument(
+            data=JsonApiDataSourcePatch(id=data_source_id, attributes=JsonApiDataSourcePatchAttributes(**attributes))
+        )
+
+
+class PostgresAttributes:
+    def __init__(self, host: str, db_name: str, port: int = 5432):
+        self.host = host
+        self.port = port
+        self.db_name = db_name
+
+    @property
+    def attributes(self) -> dict:
```

Comment:
Either use `dict[str, Any]` for return type or change all attributes to `str` as it is later serialized to URL anyway and method name can be something like `str_attributes`